packages/integration-sdk-runtime/src/execution/uploader.ts

```diff
@@ -0,0 +1,110 @@
+import { IntegrationError } from '@jupiterone/integration-sdk-core';
+import PQueue from 'p-queue/dist';
+import { FlushedGraphObjectData } from '../storage/types';
+import {
+  uploadGraphObjectData,
+  SynchronizationJobContext,
+} from '../synchronization';
+
+export interface StepGraphObjectDataUploader {
+  stepId: string;
+  enqueue: (graphObjectData: FlushedGraphObjectData) => Promise<void>;
+  waitUntilUploadsComplete: () => Promise<void>;
+}
+
+export type CreateStepGraphObjectDataUploaderFunction = (
+  stepId: string,
+) => StepGraphObjectDataUploader;
+
+export interface CreateQueuedStepGraphObjectDataUploaderParams {
+  stepId: string;
+  uploadConcurrency: number;
+  upload: (graphObjectData: FlushedGraphObjectData) => Promise<void>;
+  onThrottleEnqueue?: () => void;
+}
+
+export function createQueuedStepGraphObjectDataUploader({
+  stepId,
+  uploadConcurrency,
+  upload,
+  onThrottleEnqueue,
+}: CreateQueuedStepGraphObjectDataUploaderParams): StepGraphObjectDataUploader {
+  const queue = new PQueue({
+    concurrency: uploadConcurrency,
+  });
+
+  const uploadErrors: Error[] = [];
+
+  return {
+    stepId,
+    async enqueue(graphObjectData) {
+      if (queue.isPaused) {
+        // This step already failed an upload. We do not want to enqueue more
+        // for this step.
+        return;
+      }
+
+      // OPTIMIZATION: We do not want to buffer a lot of graph objects
+      // into memory inside of the queue. If the queue concurrency has been
+      // reached, we wait for the queue to flush so that this step has the
+      // opportunity to upload more data.
+      if (
+        queue.pending >= uploadConcurrency ||
+        queue.size >= uploadConcurrency
+      ) {
+        if (onThrottleEnqueue) {
+          // Mainly just used for testing that our custom throttling works.
+          onThrottleEnqueue();
+        }
+
+        await queue.onIdle();
```

Comment:
I think we don't want to wait for onIdle here, as it loses the concurrency benefits. Basically we won't start on the next batch of uploads until every one from the last group is finished. We should likely wait on `onEmpty` and instead of checking if wither the pending or the size is greater than the upload concurrency, I think we should just be checking the `queue.size`